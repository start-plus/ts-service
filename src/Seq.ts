export class Seq {
  private id = 0;

  getNext() {
    return ++this.id;
  }

  reset() {
    this.id = 0;
  }
}
