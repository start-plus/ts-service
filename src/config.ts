import * as Logger from 'bunyan';

export interface Config {
  removeFields: string[];
  debug: boolean;
  depth: number;
  maxArrayLength: number;
  loggerFactory: (serviceName: string) => Logger;
}

export function getDefaultConfig(): Config {
  return {
    removeFields: ['password', 'token', 'accessToken'],
    debug: true,
    depth: 4,
    maxArrayLength: 30,
    loggerFactory(serviceName) {
      return Logger.createLogger({
        name: serviceName,
        level: this.debug ? 'debug' : 'error',
      });
    },
  };
}

export const globalConfig = getDefaultConfig();

export function configure(config: Partial<Config>) {
  Object.assign(globalConfig, config);
}
