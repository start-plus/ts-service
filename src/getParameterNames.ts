/*
 From
 https://github.com/goatslacker/get-parameter-names
 and 
 https://www.npmjs.com/package/@avejidah/get-parameter-names
 */

var COMMENTS = /((\/\/.*$)|(\/\*[\s\S]*?\*\/))/gm;
var DEFAULT_PARAMS = /=[^,]+/gm;
var FAT_ARROWS = /=>.*$/gm;
var SPACES = /\s/gm;
var BEFORE_OPENING_PAREN = /^[^(]*\(/gm;
var AFTER_CLOSING_PAREN = /^([^)]*)\).*$/gm;

function joinDestructors(args: string[]) {
  let count = 0;
  const ret: string[] = [];
  let current: string[] = [];
  args.forEach(param => {
    if (param.includes('{')) {
      count++;
      current.push(param);
    } else if (param.includes('}')) {
      count--;
      current.push(param);
      if (!count) {
        ret.push(current.join(', '));
        current = [];
      }
    } else {
      ret.push(param);
    }
  });
  return ret;
}

export function getParameterNames(fn: Function) {
  var code = fn
    .toString()
    .replace(SPACES, '')
    .replace(COMMENTS, '')
    .replace(FAT_ARROWS, '')
    .replace(DEFAULT_PARAMS, '')
    .replace(BEFORE_OPENING_PAREN, '')
    .replace(AFTER_CLOSING_PAREN, '$1');

  return joinDestructors(code ? code.split(',') : []);
}
