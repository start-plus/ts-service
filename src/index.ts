export * from './decorators';
export * from './config';
