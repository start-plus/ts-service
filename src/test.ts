import * as Joi from 'joi';
import {} from 'class-validator';
import 'reflect-metadata';

const fieldMetadataKey = Symbol('field');
const schemaMetadataKey = Symbol('schema');

function Field(a: any) {
  console.log('Field(): evaluated');
  return function(
    target: any,
    propertyKey: string,
    // descriptor: PropertyDescriptor,
  ) {
    console.log('Field(): called');
    // console.log({ target, propertyKey, FooOptions });
    // Reflect.defineMetadata(fieldMetadataKey, a, target.constructor, propertyKey);
    const meta =
      Reflect.getMetadata(
        schemaMetadataKey,
        target.constructor,
        schemaMetadataKey,
      ) || {};
    meta[propertyKey] = a;
    Reflect.defineMetadata(
      fieldMetadataKey,
      a,
      target.constructor,
      schemaMetadataKey,
    );
    // console.log('meta field', meta);
  };
}

function Method(a?: any) {
  console.log('Field(): evaluated');
  return function(
    target: any,
    propertyKey: string,
    descriptor: PropertyDescriptor,
  ) {
    let paramtypes = Reflect.getMetadata(
      'design:paramtypes',
      target,
      propertyKey,
    );
    console.log('Method', { target, propertyKey });
    console.log(paramtypes);
    // const meta = Reflect.getMetadata(fieldMetadataKey, FooOptions, 'name');
    // console.log('type.name', type.name);
    // console.log('meta', meta);
    // console.log('Field(): called');
    // console.log(new FooOptions() instanceof type);
    // console.log(typeof type);
    // console.log(type.toString());
    // console.log(type().toString());
    // console.log(new (type())());
    // console.log('xxxx');
  };
}
function Options(a?: any) {
  console.log('Options(): evaluated');
  return function(
    target: Function,
    // propertyKey: string,
    // descriptor: PropertyDescriptor,
  ) {};
}

function Service(a?: any) {
  console.log('Service(): evaluated');
  return function(
    target: Function,
    // propertyKey: string,
    // descriptor: PropertyDescriptor,
  ) {
    console.log('Service(): called', target.name);
    for (const name of Object.getOwnPropertyNames(target.prototype)) {
      if (name === 'constructor') {
        continue;
      }
      const method = target.prototype[name];
      console.log(name);

      console.log('Service', { target: target.prototype, propertyKey: name });
      let paramtypes = Reflect.getOwnMetadata(
        'design:paramtypes',
        target.prototype,
        name,
      );
      console.log(paramtypes);
      // console.log(name, method);
      // console.log(Object.getOwnPropertyNames(method));
    }
    // console.log(target.prototype.foo);
  };
}

// function Param() {
//   return function(
//     target: Object,
//     propertyKey: string | symbol,
//     parameterIndex: number,
//   ) {
//     console.log('Param(): called');
//     let type = Reflect.getOwnMetadata('design:type', target, propertyKey);
//     console.log(Reflect.getMetadataKeys(target));
//     const str: any = new FooOptions();
//     console.log(str instanceof type);
//     // console.log(type('as'));
//     // for (const name of Object.getOwnPropertyNames(target.prototype)) {
//     //   if (name === 'constructor') {
//     //     continue;
//     //   }
//     //   const method = target.prototype[name];
//     //   console.log(name, method);
//     //   console.log(Object.getOwnPropertyNames(method));
//     // }
//     // console.log(target.prototype.foo);
//   };
// }

@Options()
class FooOptions {
  private constructor() {}

  @Field(Joi.string().required())
  name: string;

  @Field(Joi.number().required())
  id: number;

  @Field(Joi.object().required())
  nested: {
    address: string;
  };

  @Field(Joi.object().required())
  bar?: object;

  @Field(Joi.object().required())
  bar2: object | null;

  @Field(Joi.object().required())
  bar3: number | string;
}

const test: FooOptions = {
  name: 'foo',
  id: 123,
  nested: { address: 'aa' },
  bar2: null,
  bar3: 12,
};

class OtherOptions<T> {
  @Field(
    Joi.array()
      .items(Joi.string())
      .required(),
  )
  items: T[];
}

@Service()
class MyService {
  @Method()
  async foo(options: FooOptions) {}

  // @Method()
  // otherMethod(a: OtherOptions<string>, b: string) {
  //   return 123;
  // }
}

const f = Joi.object().required();
if (f instanceof Joi.Obj) {
}

// class Point {
//   x: number;
//   y: number;
// }

// class Line {
//   private _p0: Point;
//   private _p1: Point;

//   @validate
//   set p0(value: Point) {
//     this._p0 = value;
//   }
//   get p0() {
//     return this._p0;
//   }

//   @validate
//   set p1(value: Point) {
//     this._p1 = value;
//   }
//   get p1() {
//     return this._p1;
//   }
// }

// function validate<T>(
//   target: any,
//   propertyKey: string,
//   descriptor: TypedPropertyDescriptor<T>,
// ) {
//   let set = descriptor.set;
//   descriptor.set = function(value: T) {
//     let type = Reflect.getMetadata('design:type', target, propertyKey);
//     console.log('validate type', type);
//     if (!(value instanceof type)) {
//       throw new TypeError('Invalid type.');
//     }
//     set(value);
//   };
// }

// const l = new Line();
// l.p0 = 'abc' as any;

const def = {
  foo: {
    type: 'string',
    joi: Joi.string(),
  },
};

type Def = {
  [s: string]: {
    type: 'string' | 1;
    joi: Joi.StringSchema;
  };
};

function getType(obj: Def) {}

function guess(a: number) {
  if (a === 1) {
    return 'foo' as string;
  }
  return 100 as number;
}

const ret = guess(1);
