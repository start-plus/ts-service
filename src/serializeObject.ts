import * as util from 'util';
import { Config } from './config';

/**
 * Remove invalid properties from the object and hide long arrays
 * @param  obj the object
 * @returns the new object with removed properties
 * @private
 */
function _sanitizeObject(config: Config, obj: any) {
  if (obj === undefined) {
    return obj;
  }
  const seen: any[] = [];
  return JSON.parse(
    JSON.stringify(obj, (name, value) => {
      if (seen.indexOf(value) !== -1) {
        return '[Circular]';
      }
      if (value != null && typeof value === 'object') {
        seen.push(value);
      }
      // Array of field names that should not be logged
      // add field if necessary (password, tokens etc)
      if (config.removeFields.indexOf(name) !== -1) {
        return '<removed>';
      }
      if (name === 'req' && value && value.connection) {
        return {
          method: value.method,
          url: value.url,
          headers: value.headers,
          remoteAddress: value.connection.remoteAddress,
          remotePort: value.connection.remotePort,
        };
      }
      if (name === 'res' && value && value.statusCode) {
        return {
          statusCode: value.statusCode,
          header: value._header,
        };
      }
      if (Array.isArray(value) && value.length > config.maxArrayLength) {
        return `Array(${value.length})`;
      }
      return value;
    }),
  );
}

export function serializeObject(config: Config, obj: any) {
  return util.inspect(_sanitizeObject(config, obj), { depth: config.depth });
}
