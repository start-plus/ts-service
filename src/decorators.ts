import * as Joi from 'joi';
import * as bunyan from 'bunyan';
import * as util from 'util';
import { getParameterNames } from './getParameterNames';

const serviceMetadataKey = Symbol('service');
const schemaMetadataKey = Symbol('schema');
const validateKey = Symbol('validate');
const loggerMetadataKey = Symbol('logger');

let _seqId = 0;

const globalConfig = {
  removeFields: ['password', 'token', 'accessToken'],
  debug: true,
  depth: 4,
  maxArrayLength: 30,
  loggerFactory: (serviceName: string, config: any) =>
    bunyan.createLogger({
      name: serviceName,
      level: config.debug ? 'debug' : 'error',
    }),
};

/**
 * Remove invalid properties from the object and hide long arrays
 * @param  obj the object
 * @returns the new object with removed properties
 * @private
 */
function _sanitizeObject(obj: any) {
  if (obj === undefined) {
    return obj;
  }
  const seen: any[] = [];
  return JSON.parse(
    JSON.stringify(obj, (name, value) => {
      if (seen.indexOf(value) !== -1) {
        return '[Circular]';
      }
      if (value != null && typeof value === 'object') {
        seen.push(value);
      }
      // Array of field names that should not be logged
      // add field if necessary (password, tokens etc)
      if (globalConfig.removeFields.indexOf(name) !== -1) {
        return '<removed>';
      }
      if (name === 'req' && value && value.connection) {
        return {
          method: value.method,
          url: value.url,
          headers: value.headers,
          remoteAddress: value.connection.remoteAddress,
          remotePort: value.connection.remotePort,
        };
      }
      if (name === 'res' && value && value.statusCode) {
        return {
          statusCode: value.statusCode,
          header: value._header,
        };
      }
      if (Array.isArray(value) && value.length > globalConfig.maxArrayLength) {
        return `Array(${value.length})`;
      }
      return value;
    }),
  );
}
function _combineObject(params: any[], arr: any[]): object {
  const ret: any = {};
  arr.forEach((arg, i) => {
    ret[params[i]] = arg;
  });
  return ret;
}

function _serializeObject(obj: any) {
  return util.inspect(_sanitizeObject(obj), { depth: globalConfig.depth });
}

export function schema(joiSchema: Joi.Schema) {
  return function(target: any, propertyKey?: string, parameterIndex?: number) {
    const metadataKey = propertyKey ? parameterIndex : schemaMetadataKey;
    Reflect.defineMetadata(metadataKey, joiSchema, target, schemaMetadataKey);
  };
}

export function service(target: any) {
  const logger = globalConfig.loggerFactory(target.name, globalConfig);
  Reflect.defineMetadata(
    loggerMetadataKey,
    logger,
    target.prototype,
    loggerMetadataKey,
  );

  for (const methodName of Object.getOwnPropertyNames(target.prototype)) {
    if (methodName === 'constructor') {
      continue;
    }
    const method = target.prototype[methodName];
    const validate = Reflect.getMetadata(
      validateKey,
      target.prototype,
      methodName,
    );
    if (validate == null) {
      throw new Error(
        `Method "${
          target.name
        }.${methodName}" should be annotated with @validate or @ignore`,
      );
    }
    // if (!validate) {
    //   return;
    // }
  }
  return target;
}

export function validate(
  target: any,
  methodName: string,
  descriptor: PropertyDescriptor,
) {
  Reflect.defineMetadata(validateKey, true, target, methodName);
  const params =
    Reflect.getMetadata('design:paramtypes', target, methodName) || [];
  const method = target[methodName];
  const paramNames = getParameterNames(method);
  const keys: { [s: string]: Joi.Schema } = {};
  params.forEach((type: Function, i: number) => {
    const paramName = paramNames[i];
    const schema =
      // inline schema
      Reflect.getMetadata(i, target, schemaMetadataKey) ||
      // type schema
      Reflect.getMetadata(schemaMetadataKey, type, schemaMetadataKey);
    if (!schema) {
      throw new Error(
        `Parameter ${paramName} in ${
          target.constructor.name
        }.${methodName} is missing @schema annotation`,
      );
    }
    keys[paramName] = schema;
  });
  descriptor.value = function(...args: any[]) {
    const logger = Reflect.getMetadata(
      loggerMetadataKey,
      target,
      loggerMetadataKey,
    );
    if (!logger) {
      throw new Error(
        `Logger not configured for service ${
          target.constructor.name
        }. Make sure to add @service annotation`,
      );
    }

    const removeOutput = false; // todo
    const logExit = (output: string, id: number) => {
      const formattedOutput = removeOutput
        ? '<removed>'
        : _serializeObject(output);
      logger.debug({ id }, ` EXIT ${methodName}:`, formattedOutput);
      return output;
    };
    console.log({ params, args });
    const id = ++_seqId;
    const formattedInput = params.length
      ? _serializeObject(_combineObject(paramNames, args))
      : '{ }';
    logger.debug({ id }, `ENTER ${methodName}:`, formattedInput);
    let result;

    function withValidation(...args: any[]) {
      let normalized;
      try {
        normalized = Joi.attempt(value, schema);
      } catch (e) {
        if (method.sync) {
          throw e;
        }
        return Promise.reject(e);
      }
      const newArgs = [];
      // Joi will normalize values
      // for example string number '1' to 1
      // if schema type is number
      _.each(params, param => {
        newArgs.push(normalized[param]);
      });
      return method(...newArgs);
    }

    try {
      result = method(...args);
    } catch (e) {
      logger.error(e);
      throw e;
    }
    // promise (or async function)
    if (result && typeof result.then === 'function') {
      return result
        .then((asyncResult: any) => {
          logExit(asyncResult, id);
          return asyncResult;
        })
        .catch((e: Error) => {
          logger.error({ id }, `ERROR ${methodName}: ${formattedInput} \n`, e);
          throw e;
        });
    }
    logExit(result, id);
    return result;
  };
}

export function ignore(
  target: any,
  propertyKey: string,
  descriptor: PropertyDescriptor,
) {
  Reflect.defineMetadata(validateKey, false, target, propertyKey);
}

// @schema(Joi.object())
// class Example {
//   prop1: string;
// }

// @service
// class MyService {
//   @validate
//   searchUsers(criteria: Example) {}

//   @validate
//   getUserById(
//     @schema(Joi.string().required())
//     id: string,
//   ) {}
// }
