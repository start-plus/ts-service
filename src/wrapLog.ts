import * as Logger from 'bunyan';
import { Seq } from './Seq';
import { combineObject } from './combineObject';
import { serializeObject } from './serializeObject';
import { Config } from './config';

interface WrapLogOptions<T extends Function> {
  method: T;
  methodName: string;
  paramNames: string[];
  removeOutput: boolean;
  logger: Logger;
  seq: Seq;
  config: Config;
}

export function wrapLog<T extends Function>(options: WrapLogOptions<T>): T {
  const {
    method,
    methodName,
    paramNames,
    removeOutput,
    logger,
    seq,
    config,
  } = options;

  return (function logDecorator(...args: any[]) {
    const logExit = (output: string, id: number) => {
      const formattedOutput = removeOutput
        ? '<removed>'
        : serializeObject(config, output);
      logger.debug({ id }, ` EXIT ${methodName}:`, formattedOutput);
      return output;
    };
    const id = seq.getNext();
    const formattedInput = paramNames.length
      ? serializeObject(config, combineObject(paramNames, args))
      : '{ }';
    logger.debug({ id }, `ENTER ${methodName}:`, formattedInput);
    let result;

    try {
      result = method(...args);
    } catch (e) {
      logger.error({ id }, `ERROR ${methodName}: ${formattedInput} \n`, e);
      throw e;
    }
    // promise (or async function)
    if (result && typeof result.then === 'function') {
      return result
        .then((asyncResult: any) => {
          logExit(asyncResult, id);
          return asyncResult;
        })
        .catch((e: Error) => {
          logger.error({ id }, `ERROR ${methodName}: ${formattedInput} \n`, e);
          throw e;
        });
    }
    logExit(result, id);
    return result;
  } as any) as T;
}
