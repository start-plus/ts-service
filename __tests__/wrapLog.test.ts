import * as Logger from 'bunyan';
import { wrapLog } from '../src/wrapLog';
import { Seq } from '../src/Seq';
import { getDefaultConfig } from '../src/config';

describe('wrapLog', () => {
  function getLogger() {
    return ({
      error: jest.fn(),
      debug: jest.fn(),
    } as any) as Logger;
  }

  it('log sync', () => {
    const logger = getLogger();
    const seq = new Seq();
    const fn = wrapLog({
      logger,
      method: (a: number) => a + 10,
      methodName: 'testMethod',
      paramNames: ['a'],
      config: getDefaultConfig(),
      seq,
      removeOutput: false,
    });
    expect(fn(1)).toBe(11);
    expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
  });

  it('log async', async () => {
    const logger = getLogger();
    const seq = new Seq();
    const fn = wrapLog({
      logger,
      method: async (a: number) => a + 10,
      methodName: 'testMethod',
      paramNames: ['a'],
      config: getDefaultConfig(),
      seq,
      removeOutput: false,
    });
    await expect(fn(1)).resolves.toBe(11);
    expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
  });

  it('sanitized input', () => {
    const logger = getLogger();
    const seq = new Seq();
    const fn = wrapLog({
      logger,
      method: (
        password: string,
        values: { accessToken: string; foo: number },
      ) => 'ok',
      methodName: 'testMethod',
      paramNames: ['password', 'values'],
      config: getDefaultConfig(),
      seq,
      removeOutput: false,
    });
    expect(fn('pass', { accessToken: 'token', foo: 123 })).toBe('ok');
    expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
  });

  it('remove output', () => {
    const logger = getLogger();
    const seq = new Seq();
    const fn = wrapLog({
      logger,
      method: (a: number) => a + 10,
      methodName: 'testMethod',
      paramNames: ['a'],
      config: getDefaultConfig(),
      seq,
      removeOutput: true,
    });
    expect(fn(1)).toBe(11);
    expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
  });

  it('sync error', () => {
    const logger = getLogger();
    const seq = new Seq();
    const fn = wrapLog({
      logger,
      method: () => {
        throw new Error('some error');
      },
      methodName: 'testMethod',
      paramNames: [],
      config: getDefaultConfig(),
      seq,
      removeOutput: false,
    });
    expect(() => fn()).toThrow();
    expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
    expect((logger.error as jest.Mock<any>).mock.calls).toMatchSnapshot();
  });

  it('async error', async () => {
    const logger = getLogger();
    const seq = new Seq();
    const fn = wrapLog({
      logger,
      method: async () => {
        throw new Error('some error');
      },
      methodName: 'testMethod',
      paramNames: [],
      config: getDefaultConfig(),
      seq,
      removeOutput: false,
    });
    await expect(fn()).rejects.toThrow();
    expect((logger.debug as jest.Mock<any>).mock.calls).toMatchSnapshot();
    expect((logger.error as jest.Mock<any>).mock.calls).toMatchSnapshot();
  });
});
