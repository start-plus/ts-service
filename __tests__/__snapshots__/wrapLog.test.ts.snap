// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapLog async error 1`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER testMethod:",
    "{ }",
  ],
]
`;

exports[`wrapLog async error 2`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ERROR testMethod: { } 
",
    [Error: some error],
  ],
]
`;

exports[`wrapLog log async 1`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER testMethod:",
    "{ a: 1 }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT testMethod:",
    "11",
  ],
]
`;

exports[`wrapLog log sync 1`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER testMethod:",
    "{ a: 1 }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT testMethod:",
    "11",
  ],
]
`;

exports[`wrapLog remove output 1`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER testMethod:",
    "{ a: 1 }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT testMethod:",
    "<removed>",
  ],
]
`;

exports[`wrapLog sanitized input 1`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER testMethod:",
    "{ password: '<removed>',
  values: { accessToken: '<removed>', foo: 123 } }",
  ],
  Array [
    Object {
      "id": 1,
    },
    " EXIT testMethod:",
    "'ok'",
  ],
]
`;

exports[`wrapLog sync error 1`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ENTER testMethod:",
    "{ }",
  ],
]
`;

exports[`wrapLog sync error 2`] = `
Array [
  Array [
    Object {
      "id": 1,
    },
    "ERROR testMethod: { } 
",
    [Error: some error],
  ],
]
`;
